x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -232,10 +237,33 @@ public void restoreShard(IndexShard indexShard, SnapshotId snapshotId, Version v
         try {
             store.createEmpty(indexShard.indexSettings().getIndexMetaData().getCreationVersion().luceneVersion);
         } catch (EngineException | IOException e) {
-            throw new IndexShardRecoveryException(shardId, "failed to recover from gateway", e);
+            throw new IndexShardRecoveryException(shardId, "failed to create empty store", e);
         } finally {
             store.decRef();
         }
+
+        Store.MetadataSnapshot recoveryMetadata;
+        try {
+            recoveryMetadata = indexShard.snapshotStoreMetadata();
+        } catch (IOException e) {
+            throw new IndexShardRecoveryException(shardId, "failed access store metadata", e);
+        }
+
+        Client remoteClient = client.getRemoteClusterClient(remoteClusterAlias);
+        String sessionUUID = UUIDs.randomBase64UUID();
+        PutCcrRestoreSessionAction.PutCcrRestoreSessionResponse response = remoteClient.execute(PutCcrRestoreSessionAction.INSTANCE,
+            new PutCcrRestoreSessionRequest(sessionUUID, shardId, recoveryMetadata)).actionGet();
+        // The nodeId is necessary to route file chunk requests to appropriate node
+        String nodeId = response.getNodeId();
+
+        try {
+            // Implement file restore
+        } catch (Exception e) {
+            throw new IndexShardRecoveryException(shardId, "failed to recover from gateway", e);
+        } finally {
+            DeleteCcrRestoreSessionRequest deleteRequest = new DeleteCcrRestoreSessionRequest(sessionUUID, shardId);
+            remoteClient.execute(DeleteCcrRestoreSessionAction.INSTANCE, deleteRequest).actionGet();
```

Comment:
if this throws an exception, it might hide the original one?

Reply:
Made changes to resolve this.